File: Collections/RadixTrie.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Collections
{
    public class RadixTrie<Value> : ITrie<Value> where Value : class
    {
        public static readonly int r = 256;
        private Node root = new Node();
        private int size = 0;

        private class Node
        {
            public Value value;
            public Node[] children = new Node[r];

            public Node(Value value)
            {
                this.value = value;
            }

            public Node() { }
        }

        public Value this[string key]
        {
            get
            {
                return Get(key);
            }

            set
            {
                Put(key, value);
            }
        }

        public bool Contains(string key)
        {
            if (key == null)
                throw new ArgumentNullException();
            var node = Get(root, key, 0);
            return node != null && node.value != null;
        }

        public void Delete(string key)
        {
            if (key == null)
                throw new ArgumentNullException();

            Node Delete(Node node, int depth)
            {
                if (node == null)
                    throw new KeyNotFoundException();
                if (depth == key.Length)
                {
                    node.value = null;
                }
                else
                {
                    var c = key[depth];
                    node.children[c] = Delete(node.children[c], depth + 1);
                }

                // Optimisation: Provided we are not at the root node we may be able to get rid of this node if it carries no values
                if (depth == 0)
                {
                    return node;
                }
                else if (node.value == null && node.children.All(child => child == null))
                {
                    return null;
                }

                return node;
            }

            root = Delete(root, 0);
            size--;
        }

        public Value Get(string key)
        {
            if (key == null)
                throw new ArgumentNullException();
            var node = Get(root, key, 0);
            if (node == null || node.value == null)
                throw new KeyNotFoundException();
            return node.value;
        }

        private Node Get(Node node, string key, int depth)
        {
            if (node == null)
                return null;
            if (depth == key.Length)
                return node;  // node.value could be null if 'key' is merely a prefix of another
            return Get(node.children[key[depth]], key, depth + 1);
        }

        public Value GetOrElse(string key, Value defaultValue)
        {
            if (key == null)
                throw new ArgumentNullException();

            var node = Get(root, key, 0);
            if (node == null || node.value == null)
                return defaultValue;
            return node.value;
        }

        public bool IsEmpty() => size == 0;

        public IEnumerable<string> Keys() => Pairs().Select(p => p.key);

        public IEnumerable<string> Keys(string prefix)
        {
            if (prefix == null)
                throw new ArgumentNullException();

            var queue = new Queue<(string key, Value value)>();
            var node = Get(root, prefix, 0);  // Root of subtree for all trees begining with the given prefix
            var sb = new StringBuilder(prefix);
            Collect(node, sb, queue);
            return queue.Select(p => p.key);
        }

        public IEnumerable<(string key, Value value)> Pairs()
        {
            var queue = new Queue<(string key, Value value)>();
            var sb = new StringBuilder();
            Collect(root, sb, queue);
            return queue;
        }

        public void Put(string key, Value value)
        {
            if (key == null || value == null)
                throw new ArgumentNullException();

            Node Put(Node node, int depth)
            {
                if (node == null)
                    node = new Node();
                if (depth == key.Length)
                {
                    node.value = value;
                    return node;
                }
                var c = key[depth];
                node.children[c] = Put(node.children[c], depth + 1);
                return node;

            }

            root = Put(root, 0);
            size++;
        }

        public int Size() => size;

        public IEnumerable<Value> Values() => Pairs().Select(p => p.value);

        private static void Collect(Node node, StringBuilder sb, Queue<(string key, Value value)> queue)
        {
            if (node == null)
                return;
            if (node.value != null)
                queue.Push((sb.ToString(), node.value));
            for (int i = 0; i < r; i++)
            {
                sb.Append((char)i);
                Collect(node.children[i], sb, queue);
                sb.Length--;
            }
        }

        public string LongestPrefix(string query)
        {
            if (query == null)
                throw new ArgumentNullException();

            int LongestPrefix(Node node, int d, int l)
            {
                if (node == null)
                    return l;
                if (node.value != null)
                    l = d;
                if (d == query.Length)
                    return l;
                int c = query[d];
                return LongestPrefix(node.children[c], d + 1, l);
            }
            var length = LongestPrefix(root, 0, 0);
            return query.Substring(0, length);
        }
    }
}

File: src/main/java/in/bfdes/collections/RadixTrie.java
package in.bfdes.collections;

import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.function.BiFunction;
import java.util.function.Consumer;

public class RadixTrie<V> implements Trie<V> {
    public static final int radix = 256;  // support for ASCII chars
    private Node root = new Node();
    private int size = 0;

    private class Node {
        public V value;
        public Node[] children = (Node[]) new Object[radix];

        public Node(V value) {
            this.value = value;
        }

        public Node() {
        }
    }

    @Override
    public void put(String key, V value) {
        if (key == null || value == null)
            throw new IllegalArgumentException();

        var put = new BiFunction<Node, Integer, Node>() {
            @Override
            public Node apply(Node node, Integer depth) {
                if (node == null)
                    node = new Node();
                if (depth < key.length()) {
                    var c = key.charAt(depth);
                    node.children[c] = apply(node.children[c], depth + 1);
                    return node;
                }
                if (node.value == null)
                    size++;
                node.value = value;
                return node;
            }
        };

        root = put.apply(root, 0);
    }

    @Override
    public V get(String key) {
        if (key == null)
            throw new IllegalArgumentException();

        var node = getNode(key);
        if (node == null || node.value == null)
            throw new NoSuchElementException();
        return node.value;
    }

    @Override
    public V getOrElse(String key, V defaultValue) {
        if (key == null)
            throw new IllegalArgumentException();

        var node = getNode(key);
        if (node == null || node.value == null)
            return defaultValue;
        return node.value;
    }

    @Override
    public void delete(String key) {
        if (key == null)
            throw new IllegalArgumentException();

        var node = root;
        var depth = 0;

        while (node != null && depth < key.length())
            node = node.children[key.charAt(depth++)];

        if (node == null || node.value == null)
            return;
        size--;
        node.value = null;
    }

    @Override
    public boolean contains(String key) {
        if (key == null)
            throw new IllegalArgumentException();

        var node = getNode(key);
        return node != null && node.value != null;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public Iterator<Pair<String, V>> iterator() {
        return items("").iterator();
    }

    @Override
    public Iterable<String> keys() {
        return keys("");
    }

    @Override
    public Iterable<V> values() {
        return () -> new Iterator<>() {
            private final Iterator<Pair<String, V>> items = iterator();

            @Override
            public boolean hasNext() {
                return items.hasNext();
            }

            @Override
            public V next() {
                if (!hasNext())
                    throw new NoSuchElementException();
                return items.next().value();
            }
        };
    }

    @Override
    public Iterable<String> keys(String prefix) {
        if (prefix == null)
            throw new IllegalArgumentException();

        return () -> new Iterator<>() {
            private final Iterator<Pair<String, V>> items = items(prefix).iterator();

            @Override
            public boolean hasNext() {
                return items.hasNext();
            }

            @Override
            public String next() {
                if (!hasNext())
                    throw new NoSuchElementException();
                return items.next().key();
            }
        };
    }

    @Override
    public String longestPrefix(String query) {
        if (query == null)
            throw new IllegalArgumentException();

        var length = 0;
        var depth = 0;
        var node = root;

        while (node != null) {
            if (node.value != null)
                length = depth;
            if (depth == query.length())
                break;
            node = node.children[query.charAt(depth++)];
        }
        return query.substring(0, length);
    }

    private Node getNode(String prefix) {
        var node = root;
        var depth = 0;

        while (node != null && depth < prefix.length())
            node = node.children[prefix.charAt(depth++)];

        return node;
    }

    private Iterable<Pair<String, V>> items(String prefix) {
        var queue = new Queue<Pair<String, V>>();
        var sb = new StringBuilder(prefix);
        var node = getNode(prefix);

        var collect = new Consumer<Node>() {
            @Override
            public void accept(Node node) {
                if (node == null)
                    return;
                if (node.value != null)
                    queue.push(new Pair<>(sb.toString(), node.value));
                for (char c = 0; c < radix; c++) {
                    sb.append(c);
                    accept(node.children[c]);
                    sb.setLength(sb.length() - 1);
                }
            }
        };

        collect.accept(node);

        return queue;
    }
}
